x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/profile/RestSuggestProfilesAction.java

```diff
@@ -55,16 +66,28 @@ public String getName() {
 
     @Override
     protected RestChannelConsumer innerPrepareRequest(RestRequest request, NodeClient client) throws IOException {
-        final Set<String> dataKeys = Strings.tokenizeByCommaToSet(request.param("data", null));
         final Payload payload = request.hasContentOrSourceParam()
             ? PARSER.parse(request.contentOrSourceParamParser(), null)
-            : new Payload(null, null);
+            : new Payload(null, null, null, null);
+
+        final String data = request.param("data", null);
```

Comment:
Having `data` as a query parameter is for consistency with the GetProfileApi which does not support request body. Having it in the request body is to support having one place to specify all necessary parameters. I think each has its own usages. But we can choose to keep just one if you feel strong about it.

Conceptually, `source` and request body are really the same thing. I don't think you can provide both `source` parameter and request body: Request body takes precedence, but any unconsumed query parameter also results into error. We accept `source` as an alternative to request body because not all HTTP client supports GET with request body. So it is more for compatibility and I don't consider it as an alternative way for specifying requests (conceptually).

Reply:
Coolio, clarification makes sense. Since `source` is more about compatibility, good with supporting `data` as a param.